README.md
OntosPy
=======

RDFLIb-based Python toolkit for inspecting ontologies on the Semantic Web.

#### Version: 
Check the ``ontospy/_version.py`` file


####Installation:
``easy_install ontospy``
or
``pip install ontospy``


####Dependencies:
- <http://www.rdflib.net/> (tested on versions 2.4 and 3.0).



Description
=======

Originally, I developed this script in order to get the hang of the Python RDFLib library (note: it was previously called OntoInspector and hosted on BitBucket). RDFLib provides a number of useful primitives for working with RDF graphs; however it lacks an API aimed at interrogating and modifying a graph based on its defined schema - aka the ontology. 

OntosPy allows you to extract all the schema information from an RDFS/OWL ontology, inspect it and use it query a corresponding knowledge base. 

The basic worflow is simple: load an ontology by instantiating the ``Ontology`` class; you get back an object that lets you interrogate the RDFS/OWL schema. That's all!

Ps: the library can be used in standalone mode too.


Examples
----------

These are some basic examples that should be enough to get you started. More extensive documentation will (hopefully) be made available soon!


####Inspecting an ontology from the command line

You can use the `ontospy` script from a terminal to print out basic info about any RDF model. 

	> python ontospy.py <ontology-uri>

More options are available; use `-h` to list them all:
	
	> python ontospy.py -h


####Loading and querying the FOAF ontology



	In [1]: from ontospy import *

	In [2]: onto = Ontology("http://xmlns.com/foaf/spec/20100809.rdf")

	In [3]: onto.toplayer
	Out[3]:
	[rdflib.URIRef('http://xmlns.com/foaf/0.1/Agent'),
	 rdflib.URIRef('http://www.w3.org/2000/01/rdf-schema#Class'),
	 rdflib.URIRef('http://www.w3.org/2004/02/skos/core#Concept'),
	 rdflib.URIRef('http://xmlns.com/foaf/0.1/Document'),
	 rdflib.URIRef('http://xmlns.com/foaf/0.1/LabelProperty'),
	 rdflib.URIRef('http://www.w3.org/2000/01/rdf-schema#Literal'),
	 rdflib.URIRef('http://www.w3.org/2000/10/swap/pim/contact#Person'),
	 rdflib.URIRef('http://xmlns.com/foaf/0.1/Project'),
	 rdflib.URIRef('http://www.w3.org/2003/01/geo/wgs84_pos#SpatialThing'),
	 rdflib.URIRef('http://www.w3.org/2002/07/owl#Thing')]

	In [4]: onto.printClassTree()
	foaf:Agent
	----foaf:Group
	----foaf:Organization
	----foaf:Person
	rdfs:Class
	http://www.w3.org/2004/02/skos/core#Concept
	foaf:Document
	----foaf:Image
	----foaf:PersonalProfileDocument
	foaf:LabelProperty
	rdfs:Literal
	http://www.w3.org/2000/10/swap/pim/contact#Person
	----foaf:Person
	foaf:Project
	http://www.w3.org/2003/01/geo/wgs84_pos#SpatialThing
	----foaf:Person
	owl:Thing
	----foaf:OnlineAccount
	--------foaf:OnlineChatAccount
	--------foaf:OnlineEcommerceAccount
	--------foaf:OnlineGamingAccount


	In [5]: document = onto.classFind("document")

	In [6]: document
	Out[6]:
	[rdflib.URIRef('http://xmlns.com/foaf/0.1/Document'),
	 rdflib.URIRef('http://xmlns.com/foaf/0.1/PersonalProfileDocument')]

	In [7]: document = document[0]

	In [8]: document
	Out[8]: rdflib.URIRef('http://xmlns.com/foaf/0.1/Document')

	In [9]: onto.classAllSubs(document)
	Out[9]:
	[rdflib.URIRef('http://xmlns.com/foaf/0.1/Image'),
	 rdflib.URIRef('http://xmlns.com/foaf/0.1/PersonalProfileDocument')]

	In [10]: onto.classAllSupers(document)
	Out[10]: []

	In [11]: onto.entityComment(document)
	Out[11]: rdflib.Literal('A document.', language=None, datatype=None)



#### Sketching a Turtle model

The library includes a little utility called `sketch.py`. 

This is a (still rather rudimentary) interactive environment aimed at facilitating the initial development of RDF models.

It is meant to be used from the command line (tip: make this file executable: `chmod +x sketch.py`) and requires you to type in RDF statements using the Turtle serialization. 

*Note*: if you install ontosPy using easy_install or pip, an  executable is automatically created and added to `usr/local/bin` (on unix-based systems). You can run it by typing `sketchonto`. 

	[lambdamusic]@here:~/code/python>sketch.py 
	Good morning. Ready to Turtle away. Type docs() for help.
	In [1]: docs()

	====Sketch v 0.2====

	add()  ==> add statements to the graph
	...........SHORTCUTS:
	...........'class' = owl:Class
	...........'sub' = rdfs:subClassOf
	...........TURTLE SYNTAX:  http://www.w3.org/TR/turtle/

	show() ==> shows the graph. Can take an OPTIONAL argument for the format.
	...........eg one of['xml', 'n3', 'turtle', 'nt', 'pretty-xml', dot']

	clear()	 ==> clears the graph
	...........all triples are removed

	omnigraffle() ==> creates a dot file and opens it with omnigraffle
	...........First you must set Omingraffle as your system default app for dot files!

	quit() ==> exit

	====Have fun!====


	In [2]: add()
	Multi-line input. Enter ### when finished.
	:person a class
	:mike a :person
	:person sub :agent
	:organization sub :agent
	:worksIn rdfs:domain :person
	:worksIn rdfs:range :organization
	:mike :worksIn :DamageInc
	:DamageInc a :organization

	In [3]: show()
	@prefix : <http://this.sketch#> .
	@prefix bibo: <http://purl.org/ontology/bibo/> .
	@prefix foaf: <http://xmlns.com/foaf/0.1/> .
	@prefix npg: <http://ns.nature.com/terms/> .
	@prefix npgg: <http://ns.nature.com/graphs/> .
	@prefix npgx: <http://ns.nature.com/extensions/> .
	@prefix owl: <http://www.w3.org/2002/07/owl#> .
	@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
	@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
	@prefix skos: <http://www.w3.org/2004/02/skos/core#> .
	@prefix xml: <http://www.w3.org/XML/1998/namespace> .
	@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .

	:mike a :person ;
	    :worksIn :DamageInc .

	:worksIn rdfs:domain :person ;
	    rdfs:range :organization .

	:DamageInc a :organization .

	:organization rdfs:subClassOf :agent .

	:person a owl:Class ;
	    rdfs:subClassOf :agent .



	In [4]: show("xml")
	<?xml version="1.0" encoding="UTF-8"?>
	<rdf:RDF
	   xmlns="http://this.sketch#"
	   xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
	   xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#"
	>
	  <rdf:Description rdf:about="http://this.sketch#mike">
	    <rdf:type rdf:resource="http://this.sketch#person"/>
	    <worksIn rdf:resource="http://this.sketch#DamageInc"/>
	  </rdf:Description>
	  <rdf:Description rdf:about="http://this.sketch#organization">
	    <rdfs:subClassOf rdf:resource="http://this.sketch#agent"/>
	  </rdf:Description>
	  <rdf:Description rdf:about="http://this.sketch#DamageInc">
	    <rdf:type rdf:resource="http://this.sketch#organization"/>
	  </rdf:Description>
	  <rdf:Description rdf:about="http://this.sketch#person">
	    <rdf:type rdf:resource="http://www.w3.org/2002/07/owl#Class"/>
	    <rdfs:subClassOf rdf:resource="http://this.sketch#agent"/>
	  </rdf:Description>
	  <rdf:Description rdf:about="http://this.sketch#worksIn">
	    <rdfs:domain rdf:resource="http://this.sketch#person"/>
	    <rdfs:range rdf:resource="http://this.sketch#organization"/>
	  </rdf:Description>
	</rdf:RDF>

	In [5]: omnigraffle()
	### saves a dot file and tries to open it with your default editor
	### if you're on a mac and have omnigraffle - that could be the one!

	In [6]: quit()


CHANGES.rst
OntosPy Changelog
=======================

**note**: this file is not being maintained


November 1, 2014
---------------------------------
- added methods to extract OWL.AnnotationProperties

October 16, 2014
---------------------------------
- installed package on Pypi
- added executable: 'sketchonto' (generated via distutils)


October 11, 2014
---------------------------------
- sketch: added shortcuts 
	- (sub == rdfs:subClassOf) (class == owl:Class)
- sketch: added commands for a default_sketch


September 19, 2014
---------------------------------
- added Sketch utility
- added a dot file export mechanism 
	- so that we can open this stuff quickly in omniGraffle
	- http://paco.to/2004/omnigraffle-and-graphviz
	- http://en.wikipedia.org/wiki/DOT_(graph_description_language)

EG with labels

digraph graphname {
     a -> b [label=instanceOf];
     b -> d [label=isA];
 }


September 3, 2014
---------------------------------
- refactored ontology loading routine 


July 10, 2014
---------------------------------
- added webViz method that integrates d3 stuff
	- to be tested more!
	- added the tree view template too
		- but data format isn't finished yet (ps check also http://bl.ocks.org/mbostock/4063570)	 http://bl.ocks.org/mbostock/4339184		


2013-09-23
- added loadTriples method; separate loadURI from __setupOntology()

2013-06-02
- refactored, fixed a few bugs

2013-05-31
- added methods for proputils.pyerties

2013-05-27
- fixed ontoTree so to include OWL.Thing
- added alpha sorting to __buildClassTree
- split ontologyURI() and ontologyPrettyURI
- modified __ontologyURI so to include DC.identifier metadata
- updated ontologyAnnotations method: now multiple annotations are returned correctly
- added ontologyPhysicalLocation property

2013-05-24
- addded propertyRepresentation method
- fixed <classAllSupers> and <classAllSubs>: added wrapper so to preserve tree order
- added classRangeFor and classDomainFor; classProperties is now more generic;
- changed entityComment and entityLabel to pull out all results by default
- supers and subs methods: added parameter for sorting so to preserve tree order
- changed ontologyURI method to private; ignoring blank nodes now
- added the public -ontologyClassTree- property (previously called __classTree)
- added the FAMOUS ONTOLOGIES variables for loading stuff more easily
	eg: o = ontospy.Ontology(ontospy.FAMOUS_ONTOLOGIES.FRBR)


2013-05-09
- changed the default verbose option 


26 March 2013

- added inheritance to spy.classProperties
- improved compare script


25 March 2013 

- changed names of methods.. now all camelcased and more intuitive
- added module to compare two ontologies


sketch.py
#!/usr/bin/env python

# encoding: utf-8

"""
TURTLE SKETCH
Copyright (c) 2014 __Michele Pasin__ <michelepasin.org>. All rights reserved.

Use the interpreter interactively to create a turtle model 

Tip: make this file executable: chmod +x sketch.py


"""



import sys, os, urllib2
import logging

import rdflib	 # so we have it available as a namespace
from rdflib import exceptions, URIRef, RDFS, RDF, BNode, OWL
from rdflib.namespace import Namespace, NamespaceManager
from rdflib.plugins import sparql

# http://stackoverflow.com/questions/17393664/no-handlers-could-be-found-for-logger-rdflib-term
logging.basicConfig()




class Sketch(object):
	"""
	====Sketch v 0.2====
	
	add()  ==> add statements to the graph
	...........SHORTCUTS: 
	...........'class' = owl:Class
	...........'sub' = rdfs:subClassOf
	...........TURTLE SYNTAX:  http://www.w3.org/TR/turtle/
	
	show() ==> shows the graph. Can take an OPTIONAL argument for the format.
	...........eg one of['xml', 'n3', 'turtle', 'nt', 'pretty-xml', dot'] 
	
	clear()	 ==> clears the graph
	...........all triples are removed
	
	omnigraffle() ==> creates a dot file and opens it with omnigraffle
	...........First you must set Omingraffle as your system default app for dot files!
	
	quit() ==> exit 

	====Have fun!====
	"""
	def __init__(self, text=""):
		super(Sketch, self).__init__()
		
		self.rdfGraph = rdflib.Graph()
		self.namespace_manager = NamespaceManager(self.rdfGraph)
		
		self.SUPPORTED_FORMATS = ['xml', 'n3', 'turtle', 'nt', 'pretty-xml', 'dot']
		
		PREFIXES = [
					("", "http://this.sketch#"),
					("rdf", "http://www.w3.org/1999/02/22-rdf-syntax-ns#"),
					("rdfs", "http://www.w3.org/2000/01/rdf-schema#"),
					("xml", "http://www.w3.org/XML/1998/namespace"),
					("xsd", "http://www.w3.org/2001/XMLSchema#"),
					('foaf', "http://xmlns.com/foaf/0.1/"),
					("npg", "http://ns.nature.com/terms/"),
					("npgg", "http://ns.nature.com/graphs/"),
					("npgx", "http://ns.nature.com/extensions/"),
					("bibo", "http://purl.org/ontology/bibo/"),
					("skos", "http://www.w3.org/2004/02/skos/core#"),
					("owl", "http://www.w3.org/2002/07/owl#"),
					]
		for pref in PREFIXES:
			self.bind(pref)
		if text:
			self.add(text)
		
	def add(self, text=""):
		"""add some turtle text"""
		if not text:
			self.continuousAdd()
		else:
			pprefix = ""
			for x,y in self.rdfGraph.namespaces():
				pprefix += "@prefix %s: <%s> . \n" % (x, y)
			# add final . if missing
			if not text.strip().endswith("."):
				text += " ."
			# smart replacements
			text = text.replace(" sub ", " rdfs:subClassOf ")
			text = text.replace(" class ", " owl:Class ")
			# finally
			self.rdfGraph.parse(data=pprefix+text, format="turtle")
	
	
	def continuousAdd(self):
		print "Multi-line input. Enter ### when finished."
		temp = ""
		sentinel = "###"
		for line in iter(raw_input, sentinel):
			if line.strip() == sentinel:
				break
			if not line.strip().endswith("."):
				line += " ."	
			temp += "%s" % line
		self.add(temp)	
	
	def bind(self, prefixTuple):
		p, k = prefixTuple
		self.rdfGraph.bind(p, k)
	
	def clear(self):
		""""
		Clears the graph 
			@todo add ability to remove specific triples
		"""
		self.rdfGraph.remove((None, None, None))


	def serialize(self, aformat="turtle"):
		"""
		Serialize graph using the format required
		"""
		if aformat and aformat not in self.SUPPORTED_FORMATS:
			return "Sorry. Allowed formats are %s" % str(self.SUPPORTED_FORMATS)
		if aformat == "dot":
			return self.__serializedDot()
		else:
			# use stardard rdf serializations
			return self.rdfGraph.serialize(format=aformat)

	def __serializedDot(self):
		"""
		DOT format:
		digraph graphname {
			 a -> b [label=instanceOf];
			 b -> d [label=isA];
		 }	
		"""
		temp = ""
		for x,y,z in self.rdfGraph.triples((None, None, None)):
			temp += """"%s" -> "%s" [label="%s"];\n""" % (self.namespace_manager.normalizeUri(x), self.namespace_manager.normalizeUri(z), self.namespace_manager.normalizeUri(y))
		temp = "digraph graphname {\n%s}" % temp
		return temp


	def omnigraffle(self):
		""" tries to open an export directly in omnigraffle """
		temp = self.serialize("dot")
		
		try:  # try to put in the user/tmp folder 
			from os.path import expanduser
			home = expanduser("~")
			filename = home + "/tmp/ontospy_sketch.dot"
			f = open(filename, "w")
		except:
			filename = "ontospy_sketch.dot"
			f = open(filename, "w")
		f.write(temp)
		f.close()
		try:
			os.system("open " + filename)
		except:
			os.system("start " + filename)

		
	def show(self, aformat="turtle"):
		print self.serialize(aformat)

	def docs(self):
		print self.__docs__		









##################
# 
#  Standalone Mode:
#
##################




def main(argv):
	"""
	September 18, 2014: if an arg is passed, we visualize it
	Otherwise a simple shell gets opened. 

	"""
	
	if argv:
		print "Argument passing not implemented yet"
		if False:
			onto = Model(argv[0])
			for x in onto.getClasses():
				print x
			onto.buildPythonClasses()
			s = Sketch()
	
	else:

		intro = """Good morning. Ready to Turtle away. Type docs() for help.""" 
		# idea: every time provide a different ontology maxim!
		
		def docs():
			print "\n".join([x.strip() for x in Sketch.__doc__.splitlines()])
		
		default_sketch = Sketch()
		
		def add(text=""):
			default_sketch.add(text)
		def show(aformat=None):
			if aformat:
				default_sketch.show(aformat)
			else:
				default_sketch.show()			
		def bind(prefixTuple):
			default_sketch.bind(prefixTuple)
		def clear():
			default_sketch.clear()
		def omnigraffle():
			default_sketch.omnigraffle()
		
		
		try:
			# note: this requires IPython 0.11 or above
			import IPython
			IPython.embed(banner1=intro)
		except:
			import code
			code.interact(banner=intro, local=dict(globals(), **locals()))
	# finally
	sys.exit(0)


if __name__ == '__main__':
	main(sys.argv[1:])
